src/Builder/BuilderEncoder.php

```diff
@@ -63,110 +66,118 @@ public function encode($value): stdClass|array|string
         }
 
         return $this->wrap($value, $result);
     }
 
-    private function encodeAsObject(ExpressionInterface|StageInterface $value): stdClass
+    /**
+     * $group stage have a specific encoding because the _id argument is required and others are variadic
+     */
+    private function encodeAsGroup(GroupStage $value): stdClass
+    {
+        $result = new stdClass();
+        $result->_id = $this->recursiveEncode($value->_id);
+
+        foreach (get_object_vars($value->field) as $key => $val) {
+            $result->{$key} = $this->recursiveEncode($val);
+        }
+
+        return $this->wrap($value, $result);
+    }
+
+    private function encodeAsObject(ExpressionInterface|StageInterface|QueryInterface $value): stdClass
     {
         $result = new stdClass();
-        /** @var mixed $val */
         foreach (get_object_vars($value) as $key => $val) {
-            /** @var mixed $val */
-            $val = $this->encodeIfSupported($val);
-            if ($val !== Optional::Undefined) {
-                $result->{$key} = $val;
+            // Skip optional arguments. If they have a default value, it is resolved by the server.
+            if ($val === Optional::Undefined) {
+                continue;
             }
+
+            $result->{$key} = $this->recursiveEncode($val);
         }
 
         return $this->wrap($value, $result);
     }
 
-    private function encodeAsSingle(ExpressionInterface|StageInterface $value): stdClass
+    /**
+     * Get the unique property of the operator as value
+     */
+    private function encodeAsSingle(ExpressionInterface|StageInterface|QueryInterface $value): stdClass
     {
-        $result = [];
-        /** @var mixed $val */
         foreach (get_object_vars($value) as $val) {
-            $result = $this->encodeIfSupported($val);
-            break;
+            $result = $this->recursiveEncode($val);
+
+            // The $sum accumulator is a unary operator
+            if ($value instanceof AccumulatorInterface && is_array($result) && count($result) === 1 && array_key_first($result) === 0) {
+                $result = $result[0];
+            }
+
+            return $this->wrap($value, $result);
         }
 
-        return $this->wrap($value, $result);
+        throw new LogicException(sprintf('Class "%s" does not have a single property.', $value::class));
+    }
+
+    /**
+     * Nested arrays and objects must be encoded recursively.
+     */
+    private function recursiveEncode(mixed $value): mixed
+    {
+        if (is_array($value)) {
+            foreach ($value as $key => $val) {
+                $value[$key] = $this->recursiveEncode($val);
+            }
+
+            return $value;
+        }
+
+        if ($value instanceof stdClass) {
+            foreach (get_object_vars($value) as $key => $val) {
+                $value->{$key} = $this->recursiveEncode($val);
+            }
+        }
+
+        return $this->encodeIfSupported($value);
```

Comment:
This function is all about the `ArrayCodec` and `ObjectCodec` that were described in the codec architecture doc. 
https://github.com/mongodb/mongo-php-library/pull/1125/files#diff-8bbcff6724826a3ca2b160e0718dbee5520987bbaa22e78eda96d0d448cad5e8R27